File: android/src/main/kotlin/com/ecommpay/flutterplugin/models/PluginPaymentOptions.kt
package com.ecommpay.flutterplugin.models

import kotlinx.serialization.Serializable

@Serializable
internal data class PluginPaymentOptions(
    var actionType: PluginActionType,
    val paymentInfo: PluginPaymentInfo,
    val recurrentData: PluginRecurrentData?,
    val recipientInfo: PluginRecipientInfo?,
    val additionalFields: List<PluginAdditionalField>?,
    val screenDisplayModes: List<PluginScreenDisplayMode>?,
    val hideScanningCards: Boolean?,
    //google pay configuration
    val googleMerchantId: String?,
    val googleMerchantName: String?,
    val googleIsTestEnvironment: Boolean?,
    //theme customization
    val isDarkTheme: Boolean,
    val brandColor: String?,
    val footerLabel: String?,
    val storedCardType: Int?,
    //
    val mockModeType: PluginMockModeType
)

@Serializable
internal enum class PluginMockModeType {
    DISABLED, SUCCESS, DECLINE
}

@Serializable
internal enum class PluginActionType {
    Sale, Auth, Tokenize, Verify
}

@Serializable
internal enum class PluginScreenDisplayMode {
    HIDE_SUCCESS_FINAL_SCREEN,
    HIDE_DECLINE_FINAL_SCREEN,
    DEFAULT
}

File: ios/Classes/Models/PluginPaymentOptions.swift
//
//  PluginPaymentOptions.swift
//  ecmpplugin
//
//  Created by Alex on 18.04.2024.
//

import Foundation
import ecommpaySDK


internal enum PluginActionType: String, Decodable {
    case Sale, Auth, Tokenize, Verify
    
    func map() ->  PaymentOptions.ActionType  {
        if (self == PluginActionType.Auth) {
            return PaymentOptions.ActionType.Auth
        } else if (self == PluginActionType.Tokenize) {
            return PaymentOptions.ActionType.Tokenize
        } else if (self == PluginActionType.Verify) {
            return PaymentOptions.ActionType.Verify
        } else {
            return PaymentOptions.ActionType.Sale
        }
    }
}

internal enum PluginMockModeType: String, Decodable {
    case  DISABLED, SUCCESS, DECLINE
    
    func map() -> MockModeType  {
        if (self == PluginMockModeType.SUCCESS) {
            return MockModeType.success
        } else if (self == PluginMockModeType.DECLINE) {
            return MockModeType.decline
        } else {
            return MockModeType.disabled
        }
    }
}

internal enum PluginScreenDisplayMode: String, Decodable {
    case HIDE_SUCCESS_FINAL_SCREEN, HIDE_DECLINE_FINAL_SCREEN
    
    func map() -> ScreenDisplayMode  {
        if (self == PluginScreenDisplayMode.HIDE_SUCCESS_FINAL_SCREEN) {
            return ScreenDisplayMode.hideDeclineFinalPage
        } else {
            return ScreenDisplayMode.hideDeclineFinalPage
        }
    }
}


internal struct PluginPaymentOptions: Decodable {
    let actionType: PluginActionType
    let paymentInfo: PluginPaymentInfo
    let recurrentData: PluginRecurrentData?
    let recipientInfo: PluginRecipientInfo?
    let additionalFields: [PluginAdditionalField]?
    let screenDisplayModes: [PluginScreenDisplayMode]?
    let mockModeType: PluginMockModeType
    let hideScanningCards: Bool?
    let applePayMerchantId: String?
    let applePayDescription: String?
    let applePayCountryCode: String?
    let isDarkTheme: Bool?
    let brandColor: String?
    let storedCardType: Int?
}
